### DISCUSSION TITLE
Prisma enums not getting tree shaken in mono repo package

### DISCUSSION THREAD

**Discussion by zwhitchcox** [remix-run/remix#7737] [Q&A]
So, I'm wanting to share the authentication logic among multiple services with this code in my mono repo:

```
import type { User } from "db";
import { Role } from "db";

export function isSuper(user: User) {
  return builtinSupers.includes(user?.email || "") || user?.role === Role.Super;
}

export function checkIsAdmin(user: User) {
  return isSuper(user) || user?.role === Role.Admin;
}

export const builtinSupers = ["***@gmail.com", "***@gmail.com"];
```

Here, `Role` is an enum which contains different roles someone at the company can have. If I have this logic in my app, it works fine. I can use the `Role` enum, because it contains no server logic.

However if I import this function from my mono repo package, it bundles the prisma server logic as well:

```js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.builtinSupers = exports.checkIsAdmin = exports.isSuper = void 0;
const db_1 = require("db");
function isSuper(user) {
    return exports.builtinSupers.includes((user === null || user === void 0 ? void 0 : user.email) || "") || (user === null || user === void 0 ? void 0 : user.role) === db_1.Role.Super;
}
exports.isSuper = isSuper;
function checkIsAdmin(user) {
    return isSuper(user) || (user === null || user === void 0 ? void 0 : user.role) === db_1.Role.Admin;
}
exports.checkIsAdmin = checkIsAdmin;
exports.builtinSupers = ["***@gmail.com", "***@gmail.com"];
```

I'm not understanding why the server logic is tree shaken within the app, but not in the mono repo package. Maybe because webpack is bundling the code in the app, and `tsc` transpiles the code, removing the es modules, so webpack can't tree shake it?

Asking this question has made me decide to see what will happen if I convert the repo to an esm. Will report back results (if I remember).

**Comment by zwhitchcox** [comment]
So, changing the module worked...almost.. the only problem is that `prisma` doesn't yet support esm (https://github.com/prisma/prisma/pull/19184). However, this doesn't really matter, as all prisma code is never used on the client anyway. Except for enums.

Prisma outputs `enum` types in typescript, allowing you to have typesafety from the database to the client.

When I try use an enum in a client function, I get the builtin error:

```
✘ [ERROR] Node builtin "async_hooks" (imported by "../../pkg/db/client/runtime/library.js") must be polyfilled for the browser. You can enable this polyfill in your Remix config, e.g. `browserNodeBuiltinsPolyfill: { modules: { async_hooks: true } }` [plugin browser-node-builtins-polyfill-plugin]
```

However, this works fine when I import the prisma client from within the remix codebase. Not sure why this would be since I have both `db` in my list of `serverBundledDependencies`, and tree shaking seems to mostly work correctly. Maybe because it's an ESM module that is requiring a commonjs module?

Maybe this is something [vite](https://github.com/remix-run/remix/discussions/7632) can fix?

**Comment by zwhitchcox** [comment]
Ok, I was able to get this to work, but the process is so convoluted, it's not worth it. I would not recommend using a monorepo with Remix, at least not with prisma to share the db logic among multiple services. It's to complicated to even post. Will probably do a writeup on it later.

**Reply by tlong-dev** [reply]
I am experiencing this currently, any new tips?

**Reply by zwhitchcox** [reply]
Yes, I would change prisma client the output to a `.gitignored` directory in the monorepo db package:

```prisma
// ./prisma/schema.prisma
generator client {
  provider             = "prisma-client-js"
  previewFeatures      = ["views"]
  recursive_type_depth = "-1"
  output               = "../client"
}
```

Then, I would export two versions from the monorepo.

The browser version would just export the enums from the client:

```ts
// ./src/browser.ts
export {
	Role,
	Booking,
	Lead,
	TagState,
	Direction,
        // ...the rest of your enums  
} from "../client/edge.js";
```

while the server would export `prisma` object:

```ts
import { PrismaClient } from "@prisma/client";

let prisma: PrismaClient;

declare global {
  var __db__: PrismaClient | undefined;
}

// This is needed because in development we don't want to restart
// the server with every change, but we want to make sure we don't
// create a new connection to the DB with every change either.
// In production, we'll have a single connection to the DB.
if (process.env.NODE_ENV === "production") {
  prisma = new PrismaClient();
} else {
  if (!global.__db__) {
    global.__db__ = new PrismaClient();
  }
  prisma = global.__db__;
  prisma.$connect();
}

export { prisma };
export * from '../client'
```

This isn't the exact code, as I've taken away the things specifically for my site.

I also had to convert it to an ESM. I don't actually remember why I had to convert it to ESM, so this might not be necessary, but I do not regret it, as the new vite build system does that anyway, and it made it easier to port, but you need to add this to your `remix.config.mjs`

```js
...
	serverDependenciesToBundle: [
		/^db.*/,
...
```

Ultimately, I switched to monolithic though, and that has made things a lot simpler.

I just have both server code bases in the `app` directory, and so far, I don't see any downsides. This would be my recommendation.

**Reply by zwhitchcox** [reply]
Also, for reference, this is the `Dockerfile` I used for the monorepo:

```Dockerfile
# base node image
FROM node:lts as base
ENV PNPM_HOME="/pnpm"
ENV PATH="$PNPM_HOME:$PATH"
RUN corepack enable
ENV TZ=America/New_York

# Update and install system packages
RUN apt-get update && apt-get install -y openssl


COPY . /app
WORKDIR /app

FROM base AS build
RUN --mount=type=cache,id=pnpm,target=/pnpm/store pnpm install --frozen-lockfile
RUN pnpm run -r build

FROM build AS ui
ENV NODE_ENV production
WORKDIR /app/svc/ui
ENTRYPOINT [ "./start.sh" ]

FROM build AS bg
ENV NODE_ENV production
WORKDIR /app/svc/bg
ENTRYPOINT [ "./start.sh" ]
```

This is much simpler/faster than having separate files for the two services.

**Reply by zwhitchcox** [reply]
Also, this is the `pnpm-workspace.yaml` file. It assumes you have your monorepo packages in `./pkg` and your services in `./svc`. If you're not using pnpm, I would highly recommend it. Add this to the root of your project.

```yaml
// pnpm-workspace.yaml
packages:
  - 'pkg/*'
  - 'svc/*'
```

**Reply by zwhitchcox** [reply]
@xealot

**Reply by tlong-dev** [reply]
Nice writeup! Thank you.

For the moment I just replicated the few types I needed. But once I'm done with the problem of the moment I plan to reorganize anyway. I will go back and attempt some of this, the separate Prisma module makes a lot of sense to me.

**Reply by tlong-dev** [reply]
The monolithic Docker build is also interesting! My current version (pnpm monorepo) has a few "apps" that all have discrete Dockerfiles. This could simplify quite a bit.
